src/java.base/share/classes/jdk/internal/reflect/NativeConstructorAccessorImpl.java

```diff
@@ -28,14 +28,20 @@
 import sun.reflect.misc.ReflectUtil;
 
 import java.lang.reflect.*;
+import jdk.internal.misc.Unsafe;
 
 /** Used only for the first few invocations of a Constructor;
     afterward, switches to bytecode-based implementation */
 
 class NativeConstructorAccessorImpl extends ConstructorAccessorImpl {
+    private static final Unsafe U = Unsafe.getUnsafe();
+    private static final long GENERATED_OFFSET
+        = U.objectFieldOffset(NativeConstructorAccessorImpl.class, "generated");
+
     private final Constructor<?> c;
     private DelegatingConstructorAccessorImpl parent;
     private int numInvocations;
+    private int generated;
```

Comment:
What is the reason for using an int? I remember there was a suggestion for three states but two states seems okay so curious why it was changed from boolean to int.
The restoring to 0 in the event of failure should probably be a volatile-write. Might be clearer to declare it as a volatile.

Reply:
@AlanBateman 

> What is the reason for using an int? I remember there was a suggestion for three states but two states seems okay so curious why it was changed from boolean to int.

shipilev suggested not to use sub-word CAS, so change compareAndSetBoolean to compareAndSetInt. Change field "generated" from boolean to int doesn't increase object size (no extra byte/boolean to fold into same word).

> The restoring to 0 in the event of failure should probably be a volatile-write. Might be clearer to declare it as a volatile.
"generated" field  is read once and CAS once in this method. It doesn't likely cached and no visibiliy order required.

Reply:
okay, although I think it would be clearer for maintainers to declare it as volatile because it is set with CAS.

Reply:
@AlanBateman 

volatile attribute is added. As in common case, it only read and CAS at most once. And I see no regression on my testcase.